crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -232,91 +233,204 @@ impl ConsumerTask {
 
-                    // We got this message, let's send it through
-                    if let Err(e) = self.sender.send(&consumer_group_id, msg).await {
-                        break Err(e)
+                    match consumer.split_partition_queue(&partition.0, partition.1) {
+                        Some(queue) => {
+                            let task = topic_partition_queue_consumption_loop(
+                                self.sender.clone(),
+                                partition.clone(),
+                                queue,
+                                Arc::clone(&consumer),
+                                self.consumer_group_id.clone(),
+                            );
+
+                            if let Ok(task_id) = self.task_center.spawn_child(
+                                TaskKind::Ingress,
+                                "partition-queue",
+                                None,
+                                task,
+                            ) {
+                                topic_partition_tasks.insert(partition, task_id);
+                            } else {
+                                // shutting down
+                                return;
+                            }
+                        }
+                        None => {
+                            warn!("Invalid partition {partition} given to us in rebalance, ignoring it");
+                            continue;
+                        }
                     }
-
-                    // This method tells rdkafka that we have processed this message,
-                    // so its offset can be safely committed.
-                    // rdkafka periodically commits these offsets asynchronously, with a period configurable
-                    // with auto.commit.interval.ms
-                    if let Err(e) = consumer.store_offset(&topic, partition, offset) {
-                        break Err(e.into())
+                }
+            }
+            Rebalance::Revoke(partitions) => {
+                for partition in partitions.elements() {
+                    let partition = partition.into();
+                    match topic_partition_tasks.remove(&partition)
+                {
+                    Some(task_id) => {
+                        debug!("Stopping partitioned consumer for partition {partition} due to rebalance");
+                        // The partitioned queue will not be polled again.
+                        // It might be mid-poll right now, but if so its result will not be sent anywhere.
+                        if let Some(handle) = self.task_center.cancel_task(task_id) { handle.abort() }
                     }
+                    None => warn!("Kafka informed us of a revoked partition {partition} which we had no consumer task for"),
                 }
-                _ = &mut rx => {
-                    break Ok(());
+                }
+
+                match consumer.commit_consumer_state(CommitMode::Sync) {
```

Comment:
This is committing the offset right? Maybe just write two sentences here describing why it's safe and no messages will be lost, due to how we store the offsets in the partition topic loop

Reply:
Also did you consider making this commit mode `Async`?

You say at the top you don't need to block here, but `Sync` here definitely blocks I believe. Worst case is that some of the latest offsets won't be committed, meaning when consuming that topic/partition again, some records will be deduplicated.

(Just to be clear, I'm not sure that this is necessarily the right suggestion, but definitely a lost commit of offsets won't cause lost/duplicated messages and i was trying to follow the logic of "don't block")

Reply:
yes, i think async may make sense. i added this because otherwise on rebalance i would just see that none of the messages consumed pre rebalance were commited, because the commit timer hadn't fired yet or whatever.

Reply:
agreed its just a best effort optimisation